keps/sig-api-machinery/4153-declarative-validation/README.md

```diff
@@ -0,0 +1,1801 @@
+of the targeted release**.
+
+For enhancements that make changes to code or processes/procedures in core
+Kubernetes—i.e., [kubernetes/kubernetes], we require the following Release
+Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These
+checklist items _must_ be updated for the enhancement to be released.
+-->
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+With this enhancement, Kubernetes Developers will declare validation rules using
+IDL tags in the `types.go` files that to define the Kubernetes native API types.
+For example:
+
+```go
+// staging/src/k8s.io/api/core/v1/types.go
+
+// +validationRule="!self.hostNetwork || self.containers.all(c, c.containerPort.all(cp, cp.hostPort == cp.containerPort))"
+type PodSpec struct {
+  Containers []Container `json:...`
+}
+
+type Container struct {
+  // ...
+
+  Ports []ContainerPort `json:...`
+
+  // ...
+}
+
+type ContainerPort struct {
+  // ...
+
+  // ...
+  //
+  // +minimum=1
+  // +maximum=65535
+  HostPort int32 `json:...`
+
+  // ...
+  //
+  // +minimum=1
+  // +maximum=65535
+  ContainerPort int32 `json:...`
```

Comment:
Can this be optional? Since it will "default" to 0 (an invalid value)? ;-)

Reply:
I don't think that is relevant for this example. `ContainerPort` in practice is a required field, so it has no default and must be provided by the user. `HostPort` is optional in practice, but from what you can see in the snippet it is a required field. But really optional/required is besides the point for this example.